src/serlo-editor/plugins/text/components/text-editor.tsx

```diff
@@ -182,34 +182,31 @@ export function TextEditor(props: TextEditorProps) {
         }
 
         // Create a new Slate instance on "enter" key
-        if (isHotkey('enter', event) && !isListActive) {
-          const document = selectDocument(store.getState(), id)
-          if (!document) return
-
-          const mayManipulateSiblings = selectMayManipulateSiblings(
-            store.getState(),
-            id
-          )
-          if (!mayManipulateSiblings) return
-
-          const parent = selectParent(store.getState(), id)
-          if (!parent) return
-
-          event.preventDefault()
-
-          const slicedNodes = sliceNodesAfterSelection(editor)
+        if (isHotkey(['enter', 'shift+enter'], event) && !isListActive) {
+          // prevent two consecutive empty lines
+          // wrapped in setTimeout to let slates build in function run first
           setTimeout(() => {
-            dispatch(
-              insertPluginChildAfter({
-                parent: parent.id,
-                sibling: id,
-                document: {
-                  plugin: document.plugin,
-                  state: slicedNodes || emptyDocumentFactory().value,
-                },
-              })
-            )
+            const { path } = selection.focus
+            const node = Node.get(editor, path)
+            if (!Node.string(node).length) {
+              if (path[0] > 0) {
+                const previousNode = Node.get(editor, [path[0] - 1, 0])
+                if (!Node.string(previousNode).length)
+                  editor.deleteBackward('block')
+              } else editor.deleteBackward('block')
+            }
```

Comment:
```suggestion
            // If not an empty line, do nothing
            if (Node.string(node).length) return

            // If node does not exist in editor, delete previous block
            const nodePositionInEditor = path[0]
            if (nodePositionInEditor <= 0) {
              editor.deleteBackward('block')
              return
            }

            // If previous node has no text, delete previous block
            const previousNode = Node.get(editor, [nodePositionInEditor - 1, 0])
            if (!Node.string(previousNode).length) {
              editor.deleteBackward('block')
            }
```

1. As this code is inside `setTimeout`, early exits are possible, so using them here (more performant, lower function complexity, easier to explain with comments).
2. Assigning `path[0]` to a constant. I might have given it a wrong constant name, but that would only hint to me that it needs a name, to help understand what it is.
3. Same with comments, if my comments are wrong, then it's an indicator that comments are needed 😄